lib/core/network/protocols/http+websocket.js

```diff
@@ -0,0 +1,920 @@
+      payload.write(JSON_ENDER, offset + channel.length);
+
+      // prevent buffer overwrites due to sending packets being an
+      // async method (race condition)
+      const payloadLength = offset + channel.length + JSON_ENDER.length;
+      const payloadSafeCopy = Buffer.allocUnsafe(payloadLength);
+
+      payload.copy(payloadSafeCopy, 0, 0, payloadLength);
+
+      debugWS('Publishing to channel "realtime/%s": %s', channel, payloadSafeCopy);
+      this.server.publish(`realtime/${channel}`, payloadSafeCopy, false);
+    }
+  }
+
+  notify (data) {
+    const socket = this.socketByConnectionId.get(data.connectionId);
+    debugWS('notify: %a', data);
+
+    if (!socket) {
+      return;
+    }
+
+    const payload = data.payload;
+
+    for (let i = 0; i < data.channels.length; i++) {
+      payload.room = data.channels[i];
+      this.wsSend(socket, Buffer.from(JSON.stringify(payload)));
+    }
+  }
+
+  joinChannel (channel, connectionId) {
+    debugWS('joinChannel: %s %s', channel, connectionId);
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    debugWS('Subscribing connection ID "%s" to channel "realtime/%s"', connectionId, channel);
+    socket.subscribe(`realtime/${channel}`);
+  }
+
+  leaveChannel (channel, connectionId) {
+    debugWS('leaveChannel: %s %s', channel, connectionId);
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    socket.unsubscribe(`realtime/${channel}`);
+  }
+
+  disconnect (connectionId, message = null) {
+    debugWS('[%s] forced disconnect', connectionId);
+
+    const socket = this.socketByConnectionId.get(connectionId);
+
+    if (!socket) {
+      return;
+    }
+
+    socket.end(1011, message ? Buffer.from(message) : WS_GENERIC_CLOSE_MESSAGE);
+  }
+
+  wsOnOpenHandler (socket) {
+    const ip = Buffer.from(socket.getRemoteAddressAsText()).toString();
+    const connection = new ClientConnection(this.name, [ip]);
+
+    this.entryPoint.newConnection(connection);
+    this.connectionBySocket.set(socket, connection);
+    this.socketByConnectionId.set(connection.id, socket);
+    this.backpressureBuffer.set(socket, []);
+  }
+
+  wsOnCloseHandler (socket, code, message) {
+    const connection = this.connectionBySocket.get(socket);
+
+    if (!connection) {
+      return;
+    }
+
+    if (debugWS.enabled) {
+      debugWS(
+        '[%s] received a `close` event (CODE: %d, REASON: %s)',
+        connection.id,
+        code,
+        Buffer.from(message || '').toString());
+    }
+    this.entryPoint.removeConnection(connection.id);
+    this.connectionBySocket.delete(socket);
+    this.backpressureBuffer.delete(socket);
+    this.socketByConnectionId.delete(connection.id);
+  }
+
+  wsOnMessageHandler (socket, data) {
+    const connection = this.connectionBySocket.get(socket);
+
+    if (!data || data.byteLength === 0 || !connection) {
```

Comment:
How can we receive a message from an orphan socket ?

Reply:
Indeed.